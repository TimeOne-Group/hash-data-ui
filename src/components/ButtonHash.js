import React from 'react';
import Button from '@material-ui/core/Button';
import md5 from 'js-md5';
import sha256 from 'js-sha256';
import noAccent from '../utils/noAccent';

const performFunctions = (funcs, string) => {
  funcs.forEach((func) => {
    switch (func) {
      case 'toLowerCase':
        string = string.toLowerCase();
        break;

      case 'toUpperCase':
        string = string.toUpperCase();
        break;

      case 'no accent':
        string = noAccent(string);
        break;

      default:
        break;
    }
  });

  return string;
};

function getHashedValue(name, string) {
  const hash = {
    md5,
    sha256,
  };

  return hash[name] && hash[name](string);
}

const ButtonHash = ({
  addedColumnName,
  columns,
  rows,
  fields,
  hashing,
  dispatch,
}) => {
  const encode = () => {
    if (!addedColumnName) {
      return dispatch({ error: 'Added column name is empty' });
    }

    if (fields.filter((field) => field.completed).length !== fields.length) {
      return dispatch({ error: 'No field selected' });
    }

    const newColumns = columns.filter((column) => column.field !== 'field_add');

    newColumns.push({
      field: 'field_add',
      headerName: addedColumnName,
      width: 250,
    });

    const newRows = rows.map((row) => {
      const string = fields
        .map((field) =>
          performFunctions(
            field.funcs,
            field.selected ? row[field.selected] : field.constant
          )
        )
        .join('');

      row['field_add'] = getHashedValue(hashing, string);

      return row;
    });

    const csv = rows.reduce(
      (acc, row) => {
        const entry = Object.entries(row)
          .filter((value) => value[0] !== 'id')
          .map((value) => value[1]);

        return [...acc, entry];
      },
      [newColumns.map((column) => column.headerName)]
    );

    dispatch({ columns: newColumns, rows: newRows, csv });
  };

  return (
    <Button variant="contained" color="primary" fullWidth onClick={encode}>
      Hash
    </Button>
  );
};

export default ButtonHash;
