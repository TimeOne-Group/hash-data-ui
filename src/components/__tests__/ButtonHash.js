import React from 'react';
import '@testing-library/jest-dom';
import { render, cleanup, fireEvent } from '@testing-library/react';

import ButtonHash from '../ButtonHash';

const expectedProps = {
  addedColumnName: 'Added column',
  columns: [
    {
      field: 'field_1',
      headerName: 'Program ID',
      width: 250,
    },
    {
      field: 'field_2',
      headerName: 'Program',
      width: 250,
    },
  ],
  rows: [
    {
      id: 1,
      field_1: '1',
      field_2: 'prog1',
      field_add:
        'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad',
    },
    {
      id: 2,
      field_1: '2',
      field_2: 'prog2',
      field_add:
        'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad',
    },
  ],
  fields: [
    {
      selected: '',
      type: 'constant',
      constant: 'abc',
      completed: true,
      funcs: ['no accent'],
    },
  ],
  hashing: 'sha256',
};

describe('ButtonHash component', () => {
  afterEach(() => {
    cleanup();
  });

  it('should render correctly the button hash', () => {
    const { baseElement } = render(<ButtonHash />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should dispatch an error on click if addedColumnName props is not defined ', () => {
    const dispatch = jest.fn();
    const { getByText } = render(<ButtonHash dispatch={dispatch} />);

    fireEvent.click(getByText('Hash'));

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({
      error: 'Added column name is empty',
    });
  });

  it('should dispatch an error on click if all fileds are not completed ', () => {
    const dispatch = jest.fn();
    const notCompletedFields = [...expectedProps.fields, { completed: false }];
    const { getByText } = render(
      <ButtonHash
        dispatch={dispatch}
        addedColumnName={expectedProps.addedColumnName}
        fields={notCompletedFields}
      />
    );

    fireEvent.click(getByText('Hash'));

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({ error: 'No field selected' });
  });

  it('should dispatch newColumns, newRows and csv ', () => {
    const dispatch = jest.fn();
    const { getByText } = render(
      <ButtonHash dispatch={dispatch} {...expectedProps} />
    );

    fireEvent.click(getByText('Hash'));

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({
      columns: [
        {
          field: 'field_1',
          headerName: 'Program ID',
          width: 250,
        },
        {
          field: 'field_2',
          headerName: 'Program',
          width: 250,
        },
        {
          field: 'field_add',
          headerName: 'Added column',
          width: 250,
        },
      ],
      csv: [
        ['Program ID', 'Program', 'Added column'],
        [
          '1',
          'prog1',
          'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad',
        ],
        [
          '2',
          'prog2',
          'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad',
        ],
      ],
      rows: [
        {
          field_1: '1',
          field_2: 'prog1',
          field_add:
            'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad',
          id: 1,
        },
        {
          field_1: '2',
          field_2: 'prog2',
          field_add:
            'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad',
          id: 2,
        },
      ],
    });
  });
});
