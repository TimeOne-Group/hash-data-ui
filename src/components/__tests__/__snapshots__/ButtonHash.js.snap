// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonHash component should render correctly the button hash 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Hash
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</body>
`;
